Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp

```diff
@@ -454,15 +454,49 @@ void WebAutomationSession::platformSimulateKeyboardInteraction(WebPageProxy& pag
     }
 #endif // ENABLE(WPE_PLATFORM)
 }
+
+static bool isShiftedChar(UChar keyVal)
+{
+    // https://w3c.github.io/webdriver/#dfn-shifted-character
+    return (keyVal == '~' || keyVal == '|' || keyVal == '{' || keyVal == '}' || keyVal == '<' || keyVal == '>'
+        || keyVal == '!' || keyVal == '@' || keyVal == '#' || keyVal == '$' || keyVal == '%' || keyVal == '^'
+        || keyVal == '&' || keyVal == '*' || keyVal == '(' || keyVal == ')' || keyVal == '_' || keyVal == '+'
+        || (keyVal >= 'A' && keyVal <= 'Z'));
+}
+
 void WebAutomationSession::platformSimulateKeySequence(WebPageProxy& page, const String& keySequence)
 {
     if (page.viewBackend()) {
         platformSimulateKeySequenceLibWPE(page, keySequence, m_currentModifiers);
         return;
     }
 #if ENABLE(WPE_PLATFORM)
-    for (auto codePoint : StringView(keySequence).codePoints())
-        doKeyStrokeEvent(page, true, wpe_unicode_to_keyval(codePoint), m_currentModifiers, true);
+    // https://w3c.github.io/webdriver/#dfn-dispatch-the-events-for-a-typeable-string
+    // For each char of text:
+    for (auto codePoint : StringView(keySequence).codePoints()) {
+        // Let global key state be the result of get the global key state with input state.
+        // Note: Tracked by m_currentModifiers
+        // If char is a shifted character, and the shifted state of source is false:
+        if (isShiftedChar(codePoint) && !(platformWebModifiersFromRaw(page, m_currentModifiers).contains(WebEventModifier::ShiftKey))) {
```

Comment:
For example here, if we first get the shited state into a variable, it's easier to read and we don't need more comments.

```
bool shiftIsPressed = platformWebModifiersFromRaw(page, m_currentModifiers).contains(WebEventModifier::ShiftKey);
if (isShiftedChar(codePoint) && !shiftIsPressed)
    platformSimulateKeyboardInteraction(page, KeyboardInteraction::KeyPress, VirtualKey::Shift);
else if (!isShiftedChar(codePoint) && shiftIsPressed)
    platformSimulateKeyboardInteraction(page, KeyboardInteraction::KeyRelease, VirtualKey::Shift);
```